### ISSUE TITLE
Unnest Function In BigQuery SQLAlchemy Plugin Joins Twice When Selecting Other Columns in CTE or Subquery 

### ISSUE THREAD

**Issue by mgtcardenas** [googleapis/python-bigquery-sqlalchemy#368]
From a customer support request:

**Issue summary**: This issue is closely related to [Unnest function in BigQuery SQLAlchemy plugin joins twice when selecting other columns #353](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/353). After the previous issue was resolved, queries using the UNNEST function that are not in Common Table Expressions (CTE) or subqueries work as expected. See the following minimal example:

```
from sqlalchemy import create_engine, MetaData, select, func as F, Table, Column, Integer, ARRAY

metadata = MetaData()
table = Table(
  "table1",
  metadata,
  Column("foo", Integer),
  Column("bars", ARRAY(Integer))
)

engine = create_engine("bigquery://", future=True)
q = select(table.c.foo, F.unnest(table.c.bars).column_valued("bar"))
print(q.compile(engine))
```

This outputs the following, which is expected:
```
SELECT `table1`.`foo`, `bar` 
FROM `table1`, unnest(`table1`.`bars`) AS `bar`
```

However, if we use the table as a subquery or CTE, the issue manifests again. The following code (continuing from above):

```
q = q.cte("cte")
q = select(*q.columns)
print(q.compile(engine))
```

Produces the next output:

```
WITH `cte` AS 
(SELECT `table1`.`foo` AS `foo`, `bar` 
FROM `table1`, `table1` `table1_1`, unnest(`table1_1`.`bars`) AS `bar`)
 SELECT `cte`.`foo`, `cte`.`bar` 
FROM `cte`
```
Note that `table1` is duplicated in the `FROM` clause again.

**Expected behavior**: It is expected to generate the following query instead (note the different FROM clause):
```
WITH `cte` AS 
(SELECT `table1`.`foo`, `bar` 
FROM `table1`, unnest(`table1`.`bars`) AS `bar`)
 SELECT `cte`.`foo`, `cte`.`bar` 
FROM `cte`
```

**Comment by tswast** [comment]
@kwohlfahrt thanks for the examples and the note regarding Postgres.

**Comment by lazjake** [comment]
👋 any progress on this one? sadly running into the nested cte issue 😢

**Comment by set92** [comment]
Since this look completely stagnant, someone knows some workaround it? Is possible to write the `UNNEST(table.col) AS x` as text in the FROM?

I tried using the solution of @kwohlfahrt but didn't work because I need to know the values inside the struct, so I had to use table_valued(). 
`
sel = select(func.UNNEST(table.c[self.output_attribute_id]).table_valued(*self.input_attribute_ids)).cte("cte")
_compile_query(sel)
`

But it returns
```
SELECT `anon_1`.`pp_single`, `anon_1`.`pp_total` 

FROM
  `table` `table_1`,
  UNNEST(`table_1`.`pp_metrics`) AS `anon_1`
```
which basically duplicates the data.

**Comment by jlynchMicron** [comment]
I am getting this issue when trying to unnest an array of structs, similar to this issue: https://github.com/googleapis/python-bigquery-sqlalchemy/issues/376

**Comment by evakill** [comment]
@lazjake oh hi - still an issue 🫠

**Comment by snapiri** [comment]
IMHO the original problematic code was fixed by #839, though there is still a scenario that is not handled:

When creating a statement using JOIN + UNNEST and then using it as a subquery for a new SELECT statement, it causes the table used in the UNNEST to be specified twice in the FROM statement, thus creating implicit JOIN between that table to itself.

Example code:
```from sqlalchemy import ARRAY, Column, create_engine, func, Integer, MetaData, Table, column, select

db_metadata = MetaData()
engine = create_engine("bigquery://", future=True)

column_defs1 = [Column("id", Integer, nullable=True)]
column_defs2 = [Column("ids", ARRAY(Integer), nullable=True), Column("dummy", Integer, nullable=True)]

table1 = Table("table1", db_metadata, *column_defs1)
table2 = Table("table2", db_metadata, *column_defs2)

unnested_col_name = "unnested_ids"
unnested_ids = func.unnest(table2.c.ids).alias(unnested_col_name)
unnested_id_col = column(unnested_col_name)

q = select(table1.c.id, table2.c.dummy).select_from(unnested_ids).join(table1, table1.c.id == unnested_id_col)
print('Base query\n----------')
print(str(q.compile(engine)))

print('\nNew query\n---------')
# THIS CAUSES THE ISSUE
q = select("*").select_from(q.subquery())

print(str(q.compile(engine)))```

Which prints the following:
```Base query
----------
SELECT `table1`.`id`, `table2`.`dummy`
FROM `table2`, unnest(`table2`.`ids`) AS `unnested_ids` JOIN `table1` ON `table1`.`id` = `unnested_ids`

New query
---------
SELECT *
FROM (SELECT `table1`.`id` AS `id`, `table2`.`dummy` AS `dummy`
FROM `table2`, `table2` `table2_1`, unnest(`table2_1`.`ids`) AS `unnested_ids` JOIN `table1` ON `table1`.`id` = `unnested_ids`) AS `anon_1````

**Comment by snapiri** [comment]
IMHO the original problematic code was fixed by #839, though there is still a scenario that is not handled:

When creating a statement using JOIN + UNNEST and then using it as a subquery for a new SELECT statement, it causes the table used in the UNNEST to be specified twice in the FROM statement, thus creating implicit JOIN between that table to itself.

Example code:
```from sqlalchemy import ARRAY, Column, create_engine, func, Integer, MetaData, Table, column, select

db_metadata = MetaData()
engine = create_engine("bigquery://", future=True)

column_defs1 = [Column("id", Integer, nullable=True)]
column_defs2 = [Column("ids", ARRAY(Integer), nullable=True), Column("dummy", Integer, nullable=True)]

table1 = Table("table1", db_metadata, *column_defs1)
table2 = Table("table2", db_metadata, *column_defs2)

unnested_col_name = "unnested_ids"
unnested_ids = func.unnest(table2.c.ids).alias(unnested_col_name)
unnested_id_col = column(unnested_col_name)

q = select(table1.c.id, table2.c.dummy).select_from(unnested_ids.join(table1, table1.c.id == unnested_id_col))
print('\nRaw query\n---------')
print(str(q))
print('\nCompiled query\n--------------')
print(str(q.compile(engine)))

# print('\nNew query\n---------')
# # THIS CAUSES THE ISSUE
q = select("*").select_from(q.subquery())
print('\nRaw query\n---------')
print(str(q))
print('\nCompiled query\n--------------')
compiled = q.compile(engine)
print(str(compiled))```

Which prints the following:
```Raw query
---------
SELECT table1.id, table2.dummy
FROM table2, unnest(table2.ids) AS unnested_ids JOIN table1 ON table1.id = unnested_ids

Compiled query
---------
SELECT `table1`.`id`, `table2`.`dummy`
FROM `table2`, unnest(`table2`.`ids`) AS `unnested_ids` JOIN `table1` ON `table1`.`id` = `unnested_ids`

Raw query
---------
SELECT *
FROM (SELECT table1.id AS id, table2.dummy AS dummy
FROM table2, unnest(table2.ids) AS unnested_ids JOIN table1 ON table1.id = unnested_ids) AS anon_1

Compiled query
---------
SELECT *
FROM (SELECT `table1`.`id` AS `id`, `table2`.`dummy` AS `dummy`
FROM `table2`, `table2` `table2_1`, unnest(`table2_1`.`ids`) AS `unnested_ids` JOIN `table1` ON `table1`.`id` = `unnested_ids`) AS `anon_1`1````

Thanks to @yuval-bavli for helping with pinpointing the scenario
